File: 416-partition-equal-subset-sum-分割等和子集/416-partition-equal-subset-sum-分割等和子集.java
class Solution {
    public boolean canPartition(int[] nums) {
        int sum = 0;
        int half = 0;

        for(int i : nums){
            sum += i;
        }

        if(sum % 2 != 0) return false;

        half = sum/2;
        boolean[] dp = new boolean[half + 1];
        dp[0] = true;
        for(int i = 0; i < nums.length; i++){
            for(int j = half; j >= nums[i]; j--){
                dp[j] |= dp[j - nums[i]];
            }
        }
        return dp[half];
    }
}

File: 416-partition-equal-subset-sum-分割等和子集/416-partition-equal-subset-sum-分割等和子集.js
/**
 * @param {number[]} nums
 * @return {boolean}
 */
 var canPartition = function(nums) {
    let dp = Array(10001).fill(0)
    let sum = 0
    let half = 0
    for(let num of nums){
        sum += num
    }
    // 如果和为奇数则必定不可能分为相等的两部分    
    if(sum % 2 !== 0) return false
    
    half = sum/2

    //  01背包
    for(let i = 0; i < nums.length; i++){
        for(let j = half; j >= nums[i]; j--){
            dp[j] = Math.max(dp[j], dp[j - nums[i]]+nums[i])
        }
    }
    // console.log(dp)
    if(dp[half] === half) return true
    return false
};